File: TMS-MobileApp/src/app/types/trip.ts
// types/trip.ts

import { IConvoyeur } from "./convoyeur";
import { ICustomer } from "./customer";
import { IDriver } from "./driver";
import { IOrder } from "./order";
import { ITraject } from "./traject";
import { ITruck } from "./truck";

export interface ITrip {
  id: number;
  bookingId: string;
  tripReference: string;
  estimatedDistance: number;
  estimatedDuration: number;
  estimatedStartDate: string;
  estimatedEndDate: string;
  actualStartDate?: string;
  actualEndDate?: string;
  truckId: number;
  driverId: number;
  tripStatus: TripStatus;
  trajectId?: number| null;
  traject?: ITraject;
  
  truck?: ITruck;
  driver?: IDriver;
  deliveries?: IDelivery[];
  startLocationId?: number;
  endLocationId?: number;
  updating?: boolean;
  convoyeurId?: number | null;
  convoyeur?: IConvoyeur;
  message?: string;

  createdBy: number;
  createdByName: string;
  createdAt: string;

  updatedBy?: number | null;
  updatedByName?: string;
  updatedAt?: string | null;
}

export interface IDelivery {
  id: number;
  tripId: number;
  customerId: number;
  orderId: number;
  deliveryAddress: string;
  sequence: number;
  plannedTime?: string;
  actualArrivalTime?: string;
  actualDepartureTime?: string;
  status: DeliveryStatus;
  notes?: string;
  proofOfDelivery?: string;

 
  customer?: ICustomer;
  order?: IOrder;
}

export enum TripStatus {
  Planned = 'Planned',
  Accepted = 'Accepted',
  LoadingInProgress = 'LoadingInProgress',
  DeliveryInProgress = 'DeliveryInProgress',
  Receipt = 'Receipt',
  Cancelled = 'Cancelled'
}

export enum DeliveryStatus {
  Pending = 'Pending',
  EnRoute = 'EnRoute',
  Arrived = 'Arrived',
  Delivered = 'Delivered',
  Failed = 'Failed',
  Cancelled = 'Cancelled'
}

export const TripStatusOptions = [
  { value: TripStatus.Planned, label: 'Planifié' },
  { value: TripStatus.Accepted, label: 'Accepté' },
  { value: TripStatus.LoadingInProgress, label: 'En cours de chargement' },
  { value: TripStatus.DeliveryInProgress, label: 'En cours de livraison' },
  { value: TripStatus.Receipt, label: 'Réception' },
  { value: TripStatus.Cancelled, label: 'Annulé' }
];

export const DeliveryStatusOptions = [
  { value: DeliveryStatus.Pending, label: 'En attente' },
  { value: DeliveryStatus.EnRoute, label: 'En route' },
  { value: DeliveryStatus.Arrived, label: 'Arrivé' },
  { value: DeliveryStatus.Delivered, label: 'Livré' },
  { value: DeliveryStatus.Failed, label: 'Échoué' },
  { value: DeliveryStatus.Cancelled, label: 'Annulé' }
];

export interface CreateTripDto {
  tripReference?: string;
  estimatedDistance: number;
  estimatedDuration: number;
  estimatedStartDate: string;
  estimatedEndDate: string;
  truckId: number;
  driverId: number;
  deliveries: CreateDeliveryDto[];
  trajectId?: number | null;
  convoyeurId?: number | null;

}

export interface UpdateTripDto {
  tripReference?: string;
  estimatedDistance: number;
  estimatedDuration: number;
  estimatedStartDate: string;
  estimatedEndDate: string;
  truckId: number;
  driverId: number;
  tripStatus: TripStatus;
  deliveries: CreateDeliveryDto[];
  trajectId?: number | null;
  convoyeurId?: number | null;
}

export interface CreateDeliveryDto {
  customerId: number;
  orderId: number;
  deliveryAddress: string;
  sequence: number;
  plannedTime?: string | null;
  notes?: string | null;
}

File: backend/TransportManagementSystem/Entity/Trip.cs

using System.Text.Json.Serialization;

namespace TransportManagementSystem.Entity;

public class Trip
{
    public int Id { get; set; }
    public string BookingId { get; set; }

    public string TripReference { get; set; } // Référence métier
    public decimal EstimatedDistance { get; set; } // en km
    public decimal EstimatedDuration { get; set; } // en heures
    public DateTime? ActualStartDate { get; set; } // Réel vs planifié
    public DateTime? ActualEndDate { get; set; }
    public DateTime? EstimatedStartDate { get; set; }
    public DateTime? EstimatedEndDate { get; set; }

    // GPS coordinates for trip start and end
    public double? EndLatitude { get; set; }
    public double? EndLongitude { get; set; }
    public double? StartLatitude { get; set; }
    public double? StartLongitude { get; set; }


    public int TruckId { get; set; }
    public Truck Truck { get; set; } 

    public int DriverId { get; set; }
    public Driver Driver { get; set; } 

    public TripStatus TripStatus { get; set; } = TripStatus.Planned;

    public string? Message { get; set; } 

    public ICollection<Delivery> Deliveries { get; set; } = new List<Delivery>();

    // Multi-stops support for multi-client and multi-point trips
    public ICollection<TripStop> TripStops { get; set; } = new List<TripStop>();

    public int? TrajectId { get; set; }
    public Traject? Traject { get; set; }

    public int? ConvoyeurId { get; set; }
    public Convoyeur? Convoyeur { get; set; }

    public int? CreatedById { get; set; }
    public DateTime? CreatedAt { get; set; }
    public int? UpdatedById { get; set; }
    public DateTime? UpdatedAt { get; set; }

    [JsonIgnore]
    internal User? CreatedBy { get; set; }

    [JsonIgnore]
    internal User? UpdatedBy { get; set; }

    // GPS Tracking fields
    public string? CurrentLatitude { get; set; } // Position actuelle du chauffeur
    public string? CurrentLongitude { get; set; }
    public DateTime? LastPositionUpdate { get; set; } // Dernière mise à jour GPS

    // Assignment tracking
    public bool IsAssigned { get; set; } = false;
    public DateTime? AssignedAt { get; set; }
    public DateTime? AcceptedAt { get; set; }

}

/// <summary>
/// Statut complet du voyage pour le workflow mobile
/// </summary>
public enum TripStatus
{
    Pending,              // En attente d'assignment (nouveau)
    Assigned,             // Assigné à un chauffeur (en attente d'acceptation)
    Accepted,             // Accepté par le chauffeur
    Loading,              // Chargement en cours
    InDelivery,           // En cours de livraison (sur la route)
    Arrived,              // Arrivé à destination
    Completed,            // Livraison terminée
    Cancelled,            // Annulé
    Refused,              // Refusé par le chauffeur
    // Anciens noms pour compatibilité (deprecated)
    Planned = Pending,
    LoadingInProgress = Loading,
    DeliveryInProgress = InDelivery,
    Receipt = Arrived
}